### PR TITLE
[V6] Webapp: Support "broken arrow" button to handle PR hangs

### PR DESCRIPTION
**Pull request by leemgs**
This commit is to provide "broken arrow" button to handle unexpected system
situation such as PR hangs, frequent OOM operations, and too many zombie
processes.

**Changes proposed in this PR:**

* Version 6:
1. Updated variables (e.g., id, ip, password) for code readability

* Version 5:
1. Added a security facility to contol IP address
   - Check if IP address of a system administrator
     is equal to an allowed IP address.

* Version 4:
1. Updated a html design of Broken arrow webapge.

* Version 3:
1. Improved try ... catch statement for readability
2. Fixed incorrect file name : broke_arrow_proceed.php
3. Added comments for maintenance

* Version 2:
1. Added configuration variables in json file
2. Added a routine to handle POST html tag
3. Updated check routine if id and passowr are correct
4. Added doxygtags
5. Specified php files with permission 644

* Version 1:
1. Supported "broken arrow" button

Signed-off-by: Geunsik Lim <geunsik.lim@samsung.com>

### CODE REVIEW

File: ci/taos/webapp/broken_arrow.php [hunk 1]
```diff
@@ -0,0 +1,15 @@
+<html>
+<body>
+
+<font size=22>
+System Broken Arrow Command:
+</font> 
+
+<form action="broken_arrow_process.php" method="post">
+<li>id: <input type="text" name="id"></li><br>
+<li>password: <input type="text" name="password"></li><br>
```
[the remark this answers is not in the document]
Comment (leemgs) [reply ci/taos/webapp/broken_arrow.php]: The current PR is a skeleton. This commit will be improved  at the version 2.

**Comment by leemgs** [comment]
**Changes proposed in this PR:**

* Version 2:
1. Added configuration variables in json file
2. Added a routine to handle POST html tag
3. Updated check routine if id and passowr are correct
4. Added doxygtags
5. Specified php files with permission 644

* Version 1:
1. Supported "broken arrow" button

File: ci/taos/webapp/broken_arrow.php [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+
+<ul> <li>
+Broken Arrow refers to an accidental event that involves nuclear weapons, warheads<br>
+or components which does not create the risk of nuclear war. These include unexplained<br>
+nuclear detonation, non-nuclear detonation, radioactive contamination, and public hazard.
+</li>
+</ul>
+<br><br>
+
+<form action="broken_arrow_process.php" method="post">
```
Comment (ohsewon) [inline ci/taos/webapp/broken_arrow.php]: typo 'broken_arrow_process.php' -> 'broken_arrow_proceed.php'

File: ci/taos/webapp/broken_arrow_proceed.php [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+$json_config = json_decode($string);
+
+// Get id and passowrd from json file.
+$ba_id = $json_config->broken_arrow->id;
+$ba_pass = $json_config->broken_arrow->pass;
+
+// Run the "BROKEN ARROW" command.
+if ( $id == $ba_id && $pass == $ba_pass) {
+    //$cmd = "cat $password | su -c 'shutdown -r now'";
+    $cmd = "sudo shutdown -r now";
```
Comment (ohsewon) [inline ci/taos/webapp/broken_arrow_proceed.php]: what is different with 'sudo reboot -f'?

File: ci/taos/webapp/broken_arrow_proceed.php [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+    $output = array();
+    try{
+        echo shell_exec($cmd);
+        exec($cmd, $output);
+        system($cmd, $output);
+        }Catch(Exception e){
```
Comment (ohsewon) [inline ci/taos/webapp/broken_arrow_proceed.php]: Catch -> catch, IMO, it is better that adding space between '}' and 'c'

File: ci/taos/webapp/broken_arrow_proceed.php (hunk 3 above)
Comment (leemgs) [reply ci/taos/webapp/broken_arrow_proceed.php]: difference is a security. As you know, we have to do not use the password of system account directly.

File: ci/taos/webapp/broken_arrow.php (hunk 2 above)
Comment (leemgs) [reply ci/taos/webapp/broken_arrow.php]: Nice catch.

**Comment by leemgs** [comment]
* Web UI: 

image

File: ci/taos/webapp/broken_arrow_proceed.php (hunk 4 above)
Comment (leemgs) [reply ci/taos/webapp/broken_arrow_proceed.php]: Fixed.

**Comment by leemgs** [comment]
**Changes proposed in this PR:**

* Version 3:
1. Improved try ... catch statement for readability
2. Fixed incorrect file name : broke_arrow_proceed.php
3. Added comments for maintenance

* Version 2:
1. Added configuration variables in json file
2. Added a routine to handle POST html tag
3. Updated check routine if id and passowr are correct
4. Added doxygtags
5. Specified php files with permission 644

* Version 1:
1. Supported "broken arrow" button

File: ci/taos/webapp/broken_arrow_proceed.php [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+        echo "Caught exception: ",$e->getMessage(),"\nUnable to reboot the system....\n";
+    }
+    foreach ($output as $line) {
+        print "$line<br>";
+    }
+}
+else {
+    echo ("Oooops. ID and password is incorrect. Please check it.");
+}
+?>
```
[the remark this answers is not in the document]
Comment (leemgs) [reply ci/taos/webapp/broken_arrow_proceed.php]: Fixed.

**Comment by leemgs** [comment]
**Changes proposed in this PR:**

* Version 5:
1. Added a security facility to contol IP address
   - Check if IP address of a system administrator
     is equal to an allowed IP address.

* Version 4:
1. Updated a html design of Broken arrow webapge.

* Version 3:
1. Improved try ... catch statement for readability
2. Fixed incorrect file name : broke_arrow_proceed.php
3. Added comments for maintenance

* Version 2:
1. Added configuration variables in json file
2. Added a routine to handle POST html tag
3. Updated check routine if id and passowr are correct
4. Added doxygtags
5. Specified php files with permission 644

* Version 1:
1. Supported "broken arrow" button

File: ci/taos/webapp/broken_arrow_proceed.php [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+if ($id == "" || $pass == "")
+   die ("Unable to do the broken arrow. Please type a ID and Password.")
+
+// Run the "BROKEN ARROW" command.
+if ( $id == $ba_id && $pass == $ba_pass) {
+    // Security: Do not type a password of a system account directly.
+    $cmd = "sudo shutdown -r now";
+    $output = array();
+    try{
+         echo shell_exec($cmd);
```
Comment (ohsewon) [inline ci/taos/webapp/broken_arrow_proceed.php]: need to keep indentation.

File: ci/taos/webapp/broken_arrow_proceed.php [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+
+// Get id and passowrd from json file.
+$ba_id = $json_config->broken_arrow->id;
+$ba_pass = $json_config->broken_arrow->pass;
+$ba_ip = $json_config->broken_arrow->ip;
+
+// Security: Check if IP address of a system administrator is equal to allowed IP address.
+$ipaddress = get_client_ip();
+if ( $ba_ipaddress != "" && $ba_ipaddress != $ipaddress )
+   die ("Unable to do the broken arrow. Specify IP address in the configuration file.")
```
Comment (myungjoo) [inline ci/taos/webapp/broken_arrow_proceed.php]: ```$ba_ipaddress``` ? ```$ba_ip```?

Maybe you need to turn on strict variable declaration enforcing.

File: ci/taos/webapp/broken_arrow_proceed.php (hunk 6 above)
Comment (leemgs) [reply ci/taos/webapp/broken_arrow_proceed.php]: Fixed

File: ci/taos/webapp/broken_arrow_proceed.php (hunk 7 above)
Comment (leemgs) [reply ci/taos/webapp/broken_arrow_proceed.php]: Fixed.

**Comment by leemgs** [comment]
**Changes proposed in this PR:**

* Version 6:
1. Updated variables (e.g., id, ip, password) for code readability

* Version 5:
1. Added a security facility to contol IP address
   - Check if IP address of a system administrator
     is equal to an allowed IP address.

* Version 4:
1. Updated a html design of Broken arrow webapge.

* Version 3:
1. Improved try ... catch statement for readability
2. Fixed incorrect file name : broke_arrow_proceed.php
3. Added comments for maintenance

* Version 2:
1. Added configuration variables in json file
2. Added a routine to handle POST html tag
3. Updated check routine if id and passowr are correct
4. Added doxygtags
5. Specified php files with permission 644

* Version 1:
1. Supported "broken arrow" button

**Comment by leemgs** [comment]
The local test is successfully done. This PR (v6) is ready to be merged.

**Pull request merged**